pandas/_libs/tslibs/meson.build

```diff
@@ -0,0 +1,193 @@
+# TODO: can this be removed, I wish meson copied .pyx source to the build dir automatically
+tslibs_cython_sources_list = [
+    # List of cython sources e.g. .pyx, .pxd & __init__.py
+    # Does NOT include .pxi.in
+    '__init__.py',
+    'base.pxd',
+    'base.pyx',
+    'ccalendar.pxd',
+    'ccalendar.pyx',
+    'conversion.pxd',
+    'conversion.pyx',
+    'dtypes.pyx',
+    'dtypes.pxd',
+    'fields.pyx',
+    'nattype.pyx',
+    'nattype.pxd',
+    'np_datetime.pxd',
+    'np_datetime.pyx',
+    'offsets.pxd',
+    'offsets.pyx',
+    'parsing.pxd',
+    'parsing.pyx',
+    'period.pyx',
+    'period.pxd',
+    'strptime.pyx',
+    'timedeltas.pxd',
+    'timedeltas.pyx',
+    'timestamps.pyx',
+    'timestamps.pxd',
+    'timezones.pyx',
+    'timezones.pxd',
+    'tzconversion.pxd',
+    'tzconversion.pyx',
+    'util.pxd',
+    'vectorized.pyx',
+]
+tslibs_cython_sources = {}
+
+foreach source: tslibs_cython_sources_list
+    source_pyx = configure_file(
+        input: source,
+        output: source,
+        copy: true
+    )
+    tslibs_cython_sources += {source: source_pyx}
+endforeach
+
+# TODO: build a shared library for np_datetime and np_datetime_strings
+# and use lto?
```

Comment:
Shared libraries can be tricky in python packaging, so another option is to just build it into a `static_library()` once, and link that static library to each extension that needs it. There would still be a copy of the code statically included in each extension that uses it, but you'd avoid compiling it multiple times.

Meson will handle the lto for you if you set `b_lto=true` in the default_options.

Reply:
If I specify a .c source for example twice, is meson/ninja smart enough to avoid compiling it multiple times? 

This was an issue previously for us with setuptools since in a multi-threaded compile, there was a race in writing the object file.

Reply:
From a quick test, there's no difference in generated `build.ninja` files if I add a duplicate `.c` file to an extension module list of sources, so Meson seems to de-duplicate the input files. This shouldn't happen in the first place though?

Reply:
Meson explicitly deduplicates the list of source files for a target, so you can add the same source file multiple times. IIRC this is supposed to be a protection for the case where you have if/else logic defining the file list, and there's multiple "causes" that can result in a given file being included; like this, you do not have to check whether you already added it.

That being said, this occurs per-target, and each target has a private object directory, so if you add a duplicate `.c` file one time each to multiple extensions, it *will* compile it multiple times (it often has to, because each target i.e. an `extension_module()` can have different c_args and include directories that can change the results of compilation). But, each time it compiles it, it ***is*** smart enough to write to a different output file.

> This was an issue previously for us with setuptools since in a multi-threaded compile, there was a race in writing the object file.

It's "impossible" for Meson to have races in writing output files, since each compile rule specifies the name of its output file(s) and Meson will error out during configure time if multiple compile rules have the same output file, and ask you to rename it.